dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/service/util/CpuInfo.java

```diff
@@ -345,4 +370,69 @@ public double getSystemCpuLoad() {
             }
         }
     }
+
+    private static long lastChecked = System.currentTimeMillis();
+
+    private static final long CATCH_UP_INTERVAL = 30000;
+
+    /**
+     * Timer task that queries available process cpus.
+     */
+    class IntervalTask implements Runnable {
+
+        @Override
+        public void run() {
+            long current = System.currentTimeMillis();
+            if (lastChecked - current < CATCH_UP_INTERVAL) {
+                // on restore, we can get called back successively for every missed interval
+                // avoid the penalty of checking the Runtime in these cases
+                lastChecked = current;
+                return;
+            }
+            // find available processors
+            int runtimeAvailableProcessors = Runtime.getRuntime().availableProcessors();
+            int fileSystemAvailableProcessors = getAvailableProcessorsFromFilesystem();
+
+            int newAvailableProcessors;
+            if (fileSystemAvailableProcessors <= 0 || fileSystemAvailableProcessors > runtimeAvailableProcessors) {
+                newAvailableProcessors = runtimeAvailableProcessors;
+            } else {
+                newAvailableProcessors = fileSystemAvailableProcessors;
+            }
+
+            int currentNumberOfProcessors = AVAILABLE_PROCESSORS.get();
+            if (currentNumberOfProcessors != newAvailableProcessors) {
+                if (AVAILABLE_PROCESSORS.compareAndSet(currentNumberOfProcessors, newAvailableProcessors)) {
+                    notifyListeners(newAvailableProcessors);
+                }
+            }
+            lastChecked = System.currentTimeMillis();
+        }
+
+        public void notifyListeners(int processors) {
+            synchronized (listeners) {
+                for (AvailableProcessorsListener listener : listeners) {
+                    try {
+                        listener.setAvailableProcessors(processors);
+                    } catch (Throwable t) {
+                        if (tc.isDebugEnabled())
+                            Tr.debug(tc, "Caught exception: " + t.getMessage() + ".");
+                        FFDCFilter.processException(t, getClass().getName(), "notifyListeners");
+                    }
+                }
+            }
+        }
+    }
+
+    public class CPUCount {
+        public int get() {
+            return AVAILABLE_PROCESSORS.get();
+        }
+    }
+
+    public static void resetTimer() {
```

Comment:
Should future and activeTask be made non-static and just do INSTANCE.future.cancel(false) and INSTANCE.activeTask = ...

Could just have a reset method on CpuInfo and call INSTANCE.reset.

Also are you concerned with synchronized access to future and activeTask?  Does there need to be some concurrency control with synchronization on something?

Reply:
There's only one instance of the Activator class that calls resetTimer(), and the CpuInfo instance itself is static. Since only the constructor and the resetTimer() method access future and activeTask, I don't think we need to worry about making them non-static or synchronized. Let me know if you disagree.

Reply:
I think I understand the synchronized not being needed since future isn't reset each time you run the task.  Everything else is non-static that belongs to a CPUInfo, so that is why I was thinking these should be non static as well.

Reply:
Also it is weird to initialize a static variable in a constructor.

Reply:
The pattern for this file was to have a private constructor initialized via the static INSTANCE variable. Do you propose I change that pattern?

Reply:
No.  I am saying that if the constructor is initializing fields (activeTask and future) , I usually would expect them to be a non static field like the other field initialized in the constructor.